src/protocol/ipa_prf/boolean_ops/comparison_and_subtraction_low_com.rs

```diff
@@ -0,0 +1,439 @@
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS: WeakSharedValue + CustomArray + Field,
+    XS::Element: Field + std::ops::Not<Output = XS::Element>,
+{
+    let mut carry = AdditiveShare(XS::Element::ONE, XS::Element::ONE);
+    subtraction_circuit(ctx, record_id, x, y, &mut carry).await
+}
+
+///saturated unsigned integer subtraction
+/// subtracts y from x, Output has same length as x (we dont seem to need support for different length)
+/// when y>x, it outputs 0
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+pub async fn integer_sat_sub<C, S>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<S>,
+    y: &AdditiveShare<S>,
+) -> Result<AdditiveShare<S>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<S>: IntoIterator<Item = AdditiveShare<S::Element>>,
+    S: CustomArray + Field,
+    S::Element: Field + std::ops::Not<Output = S::Element>,
+{
+    let mut carry = AdditiveShare(S::Element::ONE, S::Element::ONE);
+    let result = subtraction_circuit(
+        ctx.narrow(&Step::SaturatedSubtraction),
+        record_id,
+        x,
+        y,
+        &mut carry,
+    )
+    .await?;
+
+    //carry computes carry=(x>=y)
+    //if carry==0 {all 0 array, i.e. Array[carry]} else {result}:
+    //compute (1-carry)*Array[carry]+carry*result =carry*result
+    let sat = AdditiveShare::<S>::expand(&carry)
+        .multiply(&result, ctx.narrow(&Step::MultiplyWithCarry), record_id)
+        .await?;
+    Ok(sat)
+}
+
+///subtraction using bit subtractor
+/// subtracts y from x, Output has same length as x (carries and indices of y too large for x are ignored)
+///implementing `https://encrypto.de/papers/KSS09.pdf` from Section 3.1/3.2
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+async fn subtraction_circuit<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+    carry: &mut AdditiveShare<XS::Element>,
+) -> Result<AdditiveShare<XS>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    XS: WeakSharedValue + CustomArray,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS::Element: Field + std::ops::Not<Output = XS::Element>,
+{
+    let mut result = AdditiveShare::<XS>::ZERO;
+    for (i, v) in x.into_iter().enumerate() {
+        result.set(
+            i,
+            bit_subtractor(
+                ctx.narrow(&BitOpStep::from(i)),
+                record_id,
+                &v,
+                y.get(i).as_ref(),
+                carry,
+            )
+            .await?,
+        );
+    }
+
+    Ok(result)
+}
+
+///bit subtractor
+///implementing `https://encrypto.de/papers/KSS09.pdf` from Section 3.1/3.2
+///output = x + !(c + y)
+///update carry to carry = ( x + carry)(!(y + carry)) + carry
+/// # Errors
+/// propagates errors from multiply
```

Comment:
I don't want to replicate the code here: https://github.com/private-attribution/ipa/blob/main/src/protocol/boolean/saturating_sum.rs#L180

The current code has better comments at the very least.

Reply:
I improved the comments, I have put BA vs GF2 on the agenda for our next meeting.